include/core/federated/network/net_common.h

```diff
@@ -179,6 +155,26 @@ THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #ifndef NET_COMMON_H
 #define NET_COMMON_H
 
+/**
+ * @brief The version of the federate protocol, encoded as three bytes.
+ * 
+ * The first byte is the major version, the next byte is the minor version
+ * and the third byte is the patch version. The first numbered version is 0.1.0 (0x000100).
+ * Earlier versions (implememented in version 0.6.0 of Lingua Franca) are designated version 0.
+ * 
+ * An implementation of the RTI may support multiple versions of the protocol, but
+ * all federates that join the federation will have to use the same version.
+ * The first federate to join may either not propose a version number (in which case
+ * the RTI will use version 0). If the RTI does not support the version proposed by the
+ * first federate, then it will send MSG_TYPE_REJECT. Otherwise, it sends MSG_TYPE_ACK.
+ * Subsequent federates will have to propose matching versions or else they too will be
+ * rejected.  If the federate itself supports multiple versions, then it can attempt to
+ * join again with a lower proposed version number.  A federate uses the MSG_TYPE_FED_VERSION
+ * to propose a protocol version (this message will also carry the federation ID).
+ * A legacy federate using version 0 attempts to join with MSG_TYPE_FED_IDS instead.
+ */
+#define LF_FED_VERSION 0x000100
```

Comment:
Do we really want to encode it this way? If we want to create a script that automatically updates the version number in the code base, that script will have to translate a semver string into a hex code... We need to store the string somewhere anyway because we need to print it when the `-version` flag is used.

Reply:
There is a tradeoff here.  Either store the binary data that will be transmitted and compared, store a string that has to be parsed at runtime, or store both and risk them being different.  Are there any other alternatives?  I opted for the first option, but don't mind if you want to change it.

Reply:
I think we should parse the string at runtime if you prefer to use a binary format for the communication. I think it would be bad to leak implementation details about the implementation into other tools and have them take care of something as low-level as turning a human-readable string into a binary format. It also obfuscates the source code. My point is that in order to print the version number, you need the the human-readable string, anyway, so you might as well store it as such.

Reply:
OK, we can do that. However, I don't really get this argument. Synthesizing a string from binary data is pretty easy using printf. So is parsing a string using scanf.

Reply:
It would be nice if 1) the define in the code was human readable, so that we can look at it and see instantly what version it denotes; 2) be able to use a `sed` command to update version numbers in the code base without having to convert the desired version number to binary. Neither are true if the define is a binary hex code.

Reply:
I'm still not convinced.  This version number will have to be sent over the network and compared, so it will be needed in binary form.  So if we give it as a string, then in C code, we will have to parse the string during startup and store the binary form.  It will have to have special parsing logic to handle "alpha", "beta", "snapshot", etc., whereas in binary form these could just be encoded in the final byte.  For the sed script to update the version number, is it more difficult to replace "#define LF_REACTOR_C_VERSION 0x000600" (for 0.6.0) with "#define LF_REACTOR_C_VERSION 0x000700" than to replace strings?

Also, there is no command-line interface for reactor-c, so it's not clear that this version number will ever have to converted to a string.  One of my longer-term goals is to have a pico version of reactor-c that uses no printf (or scanf) anywhere, since these are quite expensive in program memory.

Reply:
We were discussing versioning of the rti and now we're talking about a specific variant of reactor-c for tiny embedded system. I'm a bit lost...

What I think would be reasonable requirements are:
 - a human-readable version number in the source code
 - ability to have an simple script that uses `sed` to update the version number based on a string argument, which we'll use with every release
 - ability to run **the rti** with a `--version` flag to print the version number

What I'm saying nothing about is:
 - what restrictions to impose on the version number (e.g., dot-separated numerals only)
 - how to send the version number across the network
 - how to extract a version number from a program that uses `reactor-c`
 - anything about trimmed-down runtime implementations that won't print anything in the first place

If you don't think the stated requirements are reasonable, please let me know why you think they aren't.

Reply:
You are right.  This needs more thought.  I'm going to table this for now.  Other things have higher priority.

Reply:
> This needs more thought.

Does it, really? I would just stick a `#define VERSION "0.6.0"` in `rti_common.h` and add a function in `rti_remote.c` to turn it into a binary number before using it in the protocol. Seems simple. Am I missing something?

Reply:
I've seen CMake generate a `version.h` headerfile that can be included by any target, defining such a `VERSION` macro. This way CMake can intelligently produce major or minor revisions and targets can use as they wish.

Reply:
I've shelved this PR for now because I have higher priority items to work on.  Happy to yield if someone else can take over.